proposals/p2240.md

```diff
@@ -0,0 +1,1144 @@
+// Invokes f, with the tuple `args` as its arguments.
+fn Apply[... each T:! type, F:! CallableWith(... each T)]
+    (f: F, args: (... each T)) -> auto {
+  return f(...expand args);
+}
+```
+
+```carbon
+// Takes an arbitrary number of vectors with arbitrary element types, and
+// returns a vector of tuples where the i'th element of the vector is
+// a tuple of the i'th elements of the input vectors.
+fn Zip[... each ElementType:! type]
+      (... each vector: Vector(each ElementType))
+      -> Vector((... each ElementType)) {
+  ... var each iter: auto = each vector.Begin();
+  var result: Vector((... each ElementType));
+  while (...and each iter != each vector.End()) {
+    result.push_back((... each iter));
+    ... each iter++;
+  }
+  return result;
+}
+```
+
+```carbon
+// Toy example of mixing variadic and non-variadic parameters.
+// Takes an i64, any number of f64s, and then another i64.
+fn MiddleVariadic(first: i64, ... each middle: f64, last: i64);
+```
+
+```carbon
+// Toy example of using the result of variadic type deduction.
+fn TupleConcat[... each T1: type, ... each T2: type](
+    t1: (... each T1), t2: (... each T2)) -> (... each T1, ... each T2) {
+  return (...expand t1, ...expand t2);
+}
+```
+
+### Comparisons
+
+The following table compares selected examples of Carbon variadics against
+equivalent code written in C++20 (with and without the extensions discussed
+[earlier](#background)) and Swift.
+
+<table>
+<tr>
+<th>Carbon</th>
+<th>C++20</th>
+<th>C++20 with extensions</th>
+<th>Swift with extensions</th>
+</tr>
+
+<tr>
+<td>
+
+```carbon
+// Computes the sum of its arguments, which are i64s
+fn SumInts(... each param: i64) -> i64 {
+  var sum: i64 = 0;
+  ... sum += each param;
+  return sum;
+}
+```
+
+</td>
+<td>
+
+```C++
+template <typename... Params>
+    requires (std::convertible_to<Params, int64_t> && ...)
+int64_t SumInts(const Params&... params) {
+  return (static_cast<int64_t>(params) + ... + 0);
+}
+```
+
+</td>
+<td>
+
+With P1219R2:
+
+```C++
+int64_t SumInts(int64... params) {
+  return (static_cast<int64_t>(params) + ... + 0);
+}
+```
+
+</td>
+<td>
+
+(No extensions)
+
+```swift
+func SumInts(_ params: Int64...) {
+  var sum: Int64 = 0
+  for param in params {
+    sum += param
+  }
+  return sum
+}
+```
+
+</td>
+</tr>
+<tr>
+<td>
+
+```carbon
+fn Min[T:! Comparable & Value](... each(>=1) param: T) -> T {
+  let (var result: T, ... each next: T) = (... each param);
+  ... if (each next < result) {
+    result = each next;
+  }
+  return result;
+}
+```
+
+</td>
+<td>
+
+```c++
+template <typename T>
+    requires std::totally_ordered<T> && std::copyable<T>
+T Min(const T& t) {
+  return t;
+}
+
+template <typename T, typename... Params>
+    requires std::totally_ordered<T> && std::copyable<T> &&
+             (std::same_as<T, Params> && ...)
+T Min(T first, Params... rest) {
+  T min_rest = Min(rest...);
+  if (min_rest < first) {
+    return min_rest;
+  } else {
+    return first;
+  }
+}
```

Comment:
I would put the `// Base case.` comment inside the first branch of the `if`.

I feel very mixed about this suggestion since the overloading approach was the way to do this in the day, but I guess C++ has moved and the example is documented as using C++20.

Reply:
I don't think this is an improvement for `StrCat`, where it creates two edge cases. @zygoloid had an alternate implementation that uses fold expressions to avoid quadratic compile time and the need for a helper function, but I don't know if that serves the narrative purpose you are looking for. Is this supposed to be a common way of writing this function, a way that shows common pitfalls, or a way that is as similar as possible to the Carbon approach?

Reply:
I had been aiming for these examples to be representative of how functions like this would typically be written. I think that means we shouldn't use the comma-fold trick, and probably not `if constexpr` either, although that's more debatable. However, it sounds like maybe Chandler is asking for the examples to represent the ideal versions of these functions; in that case, it might make sense to use the comma-fold trick in `StrCat`. In either case, I agree with you that `StrCat` should not use `if constexpr`, but I'll wait for @chandlerc to clarify what he'd like to see.

Reply:
In an offline conversation, Chandler indicated he thinks the C++ examples should be as elegant and optimistic as we can make them, unless we're specifically focusing on and discussing the issue that typical C++ code is suboptimal in some way. We're not focusing on that kind of issue here, so I've rewritten this to the comma-fold version.

(Chandler, feel free to clarify/correct my paraphrase of what you said).

Reply:
Looks good to me!